zkevm-circuits/src/taiko_pi_circuit.rs

```diff
@@ -27,9 +62,99 @@ const RPI_BYTES_LEN: usize = 32 * 10;
+
+const RLP_HDR_NOT_SHORT: u64 = 0x81;
+
+// Maximum size of block header fields in bytes
+const PARENT_HASH_SIZE: usize = WORD_SIZE;
+const OMMERS_HASH_SIZE: usize = WORD_SIZE;
+const BENEFICIARY_SIZE: usize = ADDRESS_SIZE;
+const STATE_ROOT_SIZE: usize = WORD_SIZE;
+const TX_ROOT_SIZE: usize = WORD_SIZE;
+const RECEIPTS_ROOT_SIZE: usize = WORD_SIZE;
+const LOGS_BLOOM_SIZE: usize = 256;
+const DIFFICULTY_SIZE: usize = 1;
+const NUMBER_SIZE: usize = U64_SIZE;
+const GAS_LIMIT_SIZE: usize = WORD_SIZE;
+const GAS_USED_SIZE: usize = WORD_SIZE;
+const TIMESTAMP_SIZE: usize = WORD_SIZE;
+const EXTRA_DATA_SIZE: usize = 1;
+const MIX_HASH_SIZE: usize = WORD_SIZE;
+const NONCE_SIZE: usize = U64_SIZE;
+const BASE_FEE_SIZE: usize = WORD_SIZE;
+const WITHDRAWALS_ROOT_SIZE: usize = WORD_SIZE;
+
+// Helper contants for the offset calculations below
+const PARENT_HASH_RLP_LEN: usize = PARENT_HASH_SIZE + 1;
+const OMMERS_HASH_RLP_LEN: usize = OMMERS_HASH_SIZE + 1;
+const BENEFICIARY_RLP_LEN: usize = BENEFICIARY_SIZE + 1;
+const STATE_ROOT_RLP_LEN: usize = STATE_ROOT_SIZE + 1;
+const TX_ROOT_RLP_LEN: usize = TX_ROOT_SIZE + 1;
+const RECEIPTS_ROOT_RLP_LEN: usize = RECEIPTS_ROOT_SIZE + 1;
+const LOGS_BLOOM_RLP_LEN: usize = LOGS_BLOOM_SIZE + 3;
+const DIFFICULTY_RLP_LEN: usize = DIFFICULTY_SIZE;
+const NUMBER_RLP_LEN: usize = NUMBER_SIZE + 1;
+const GAS_LIMIT_RLP_LEN: usize = GAS_LIMIT_SIZE + 1;
+const GAS_USED_RLP_LEN: usize = GAS_USED_SIZE + 1;
+const TIMESTAMP_RLP_LEN: usize = TIMESTAMP_SIZE + 1;
+const EXTRA_DATA_RLP_LEN: usize = EXTRA_DATA_SIZE;
+const MIX_HASH_RLP_LEN: usize = MIX_HASH_SIZE + 1;
+const NONCE_RLP_LEN: usize = NONCE_SIZE + 1;
+const BASE_FEE_RLP_LEN: usize = BASE_FEE_SIZE + 1;
+const WITHDRAWALS_ROOT_RLP_LEN: usize = WITHDRAWALS_ROOT_SIZE;
+
+// Row offsets where the value of block header fields start (after their RLP
+// header)
+const PARENT_HASH_RLP_OFFSET: usize = 4;
+const BENEFICIARY_RLP_OFFSET: usize =
+    PARENT_HASH_RLP_OFFSET + PARENT_HASH_RLP_LEN + OMMERS_HASH_RLP_LEN;
+const STATE_ROOT_RLP_OFFSET: usize = BENEFICIARY_RLP_OFFSET + BENEFICIARY_RLP_LEN;
+const TX_ROOT_RLP_OFFSET: usize = STATE_ROOT_RLP_OFFSET + STATE_ROOT_RLP_LEN;
+const RECEIPTS_ROOT_RLP_OFFSET: usize = TX_ROOT_RLP_OFFSET + TX_ROOT_RLP_LEN;
+const NUMBER_RLP_OFFSET: usize =
+    RECEIPTS_ROOT_RLP_OFFSET + RECEIPTS_ROOT_RLP_LEN + LOGS_BLOOM_RLP_LEN + DIFFICULTY_RLP_LEN;
+const GAS_LIMIT_RLP_OFFSET: usize = NUMBER_RLP_OFFSET + NUMBER_RLP_LEN;
+const GAS_USED_RLP_OFFSET: usize = GAS_LIMIT_RLP_OFFSET + GAS_LIMIT_RLP_LEN;
+const TIMESTAMP_RLP_OFFSET: usize = GAS_USED_RLP_OFFSET + GAS_USED_RLP_LEN;
+const MIX_HASH_RLP_OFFSET: usize = TIMESTAMP_RLP_OFFSET + TIMESTAMP_RLP_LEN + EXTRA_DATA_RLP_LEN;
+const BASE_FEE_RLP_OFFSET: usize = MIX_HASH_RLP_OFFSET + MIX_HASH_RLP_LEN + NONCE_RLP_LEN;
+const WITHDRAWALS_ROOT_RLP_OFFSET: usize = BASE_FEE_RLP_OFFSET + BASE_FEE_RLP_LEN;
+const BLOCKHASH_TOTAL_ROWS: usize = WITHDRAWALS_ROOT_RLP_OFFSET + WITHDRAWALS_ROOT_RLP_LEN;
```

Comment:
After assessing this, I'm not sure it is worth it.

The constants describe the block layout which is not that dynamic and since its not going to change that often I don't see much risk in errors. There was no single bug I've faced that had to do with this part of the code.

Also, I don't see any points in the code that could benefit from this change except maybe saving 10-15 LOC here https://github.com/taikoxyz/zkevm-circuits/pull/129/files#diff-55f686c17b36ef6bb92c6d15364561ab33f6f2c7ed2df3334f185a10cf49623bR1741-R1793. 

In my opinion replacing these consts with calls to functions and enum params will add more noise to the code throughout - I find the current way cleaner and more straightforward to read. 

I think in this case it is better to leave the code more explicit and simple than make it more complex for the sake of saving a few LOCs.

Reply:
Okay makes sense! I thought maybe it could make some things simpler, but if that's not the case than it's definitely not worth it.